### PR TITLE
Encode emails properly before sending to /confirm-email

### PR DESCRIPTION
**Pull request by Zhao-Andy**
## What type of PR is this? (check all applicable)
- [x] Bug Fix

## Description
Before, some emails were not encoded properly, `andy+hi@forem.com`. This PR ensures that the email is always encoded before redirecting, so that symbols like `+` are not scrubbed accidentally.

## Related Tickets & Documents
Closes https://github.com/forem/internalEngineering/issues/105

## QA Instructions, Screenshots, Recordings
1. Visit localhost:3000/confirm-email?email=some+email@test.com
2. Check that it properly displays on the page and form

### UI accessibility concerns?
Nope!

## Added tests?
- [x] Yes

## Added to documentation?
- [x] No documentation needed

## [optional] Are there any post deployment tasks we need to perform?
Nope

### CODE REVIEW

File: spec/system/authentication/user_logs_in_with_email_spec.rb [hunk 1]
```diff
@@ -13,25 +13,33 @@
...
-
-        fill_in_user(user)
-        click_button("Sign up", match: :first)
+        sign_up_user
 
         expect(page).to have_current_path("/confirm-email", ignore_query: true)
       end
+
+      it "displays the properly decoded email" do
+        decoded_email = user.email.sub("@", "+something@")
```
Comment (rhymes) [inline spec/system/authentication/user_logs_in_with_email_spec.rb:37]: `CGI.unescape` ?

**Review by rhymes — CHANGES_REQUESTED** [review]
I think the display and the input box value of the email need to encoded too:

Screenshot 2021-01-12 at 10 12 16 AM

Screenshot 2021-01-12 at 10 12 45 AM

File: spec/system/authentication/user_logs_in_with_email_spec.rb (hunk 1 above)
Comment (Zhao-Andy) [reply spec/system/authentication/user_logs_in_with_email_spec.rb:37]: Hmm sorry? `decoded_email` is to mock an email address that has the `+` in it, and `CGI.unescape` would remove that.

**Comment by Zhao-Andy** [comment]
@rhymes ah that's a good point, but I think we don't need to encode that when the page is visited directly. I know I said to try that in my directions though 😅 Sorry about that. From what I can tell, the only time a user lands on `/confirm-email` is the redirect after registering. I might be wrong though, and if I am then yeah we would need to cover the case you mentioned.

I had trouble trying to do so earlier because by the time you're visiting the page, Rails has already `CGI.unescape`d the email param, and I'm not sure how to get to the email before it's touched. 😕

**Review by msarit — APPROVED** [review]
Hey Andy! great work! I experienced the same behavior as detailed by Rhymes, but other than that, things look good to me, so I'll go ahead and approve.

File: spec/system/authentication/user_logs_in_with_email_spec.rb (hunk 1 above)
Comment (rhymes) [reply spec/system/authentication/user_logs_in_with_email_spec.rb:37]: ah sorry!

**Comment by rhymes** [comment]
@Zhao-Andy not sure I have understood. I started the app, went to `http://localhost:3000/confirm-email?email=andy+hi@forem.com` and when I pressed "Resend confirmation instructions" I got the error, as the email with the space is not valid. As that button is in `app/views/devise/confirmations/new.html.erb`, can't we just unescape the `value` attribute of the input box?

**Comment by Zhao-Andy** [comment]
@rhymes, we could, but the value is already `andy hi@forem.com` when you visit the page with unescaped URL param `?email=andy+hi@forem.com`, so encoding it would only encode the `whitespace`. Since the only place `/confirm-email` is being used is as a redirect registration, I think we only need to encode the URL param in the redirect location of `registrations_controller.rb`.

Sorry, I should update my directions to say:
1. visit `localhost:3000/confirm-email?email=andy%2Bhi%40forem.com`
2. See that the URL is displayed properly in the input and page

@msarit maybe you know since you've worked on this area, but would a user ever visit `/confirm-email` directly? Or is it only ever used as a redirect? Seems like the second from what I could find.

**Comment by rhymes** [comment]
@Zhao-Andy can they land directly there from the confirmation email?

**Comment by Zhao-Andy** [comment]
@rhymes Since we have no mailer view that has that link I assume no.

**Comment by msarit** [comment]
> @msarit maybe you know since you've worked on this area, but would a user ever visit /confirm-email directly? Or is it only ever used as a redirect? Seems like the second from what I could find.

Hey @Zhao-Andy 👋🏾  yes, from what I know, `/confirm-email` is reached via redirect normally.

**Comment by msarit** [comment]
@Zhao-Andy Is it normal that after clicking on "Resend email confirmation", I am redirected to http://localhost:3000/new?signin=true (the New Post page essentially)?

**Comment by Zhao-Andy** [comment]
@msarit I think for whatever reason the resend confirmation is sending the user to the generic "registration complete" route, which can be the homepage or `/new`. Possibly a bug but I'll leave it for a different PR.

**Review by rhymes — APPROVED** [review]
LGTM

**Review by benhalpern — APPROVED** [review]

**Pull request merged**